lang/funcs/core/coresys/cpucount_fact.go

```diff
@@ -0,0 +1,201 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package coresys
+
+import (
+	"io/ioutil"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/purpleidea/mgmt/lang/funcs/facts"
+	"github.com/purpleidea/mgmt/lang/types"
+
+	"github.com/purpleidea/mgmt/lib/socketset"
+	"golang.org/x/sys/unix"
+
+	errwrap "github.com/pkg/errors"
+)
+
+const (
+	rtmGrps         = 0x1 // make me a multicast reciever
+	socketFile      = "pipe.sock"
+	cpuDevpathRegex = "/devices/system/cpu/cpu[0-9]"
+)
+
+// CPUCountFact is a fact that returns the current CPU count
+type CPUCountFact struct {
+	init      *facts.Init
+	closeChan chan struct{}
+}
+
+func init() {
+	facts.ModuleRegister(moduleName, "cpu_count", func() facts.Fact { return &CPUCountFact{} }) // must register the fact and name
+}
+
+// Init run startup code for this fact. Initializes the closeChan
+// and sets the facts.Init variable
+func (obj *CPUCountFact) Init(init *facts.Init) error {
+	obj.init = init
+	obj.closeChan = make(chan struct{})
+	return nil
+}
+
+type nlChanEvent struct {
+	uevent *socketset.UEvent
+	err    error
+}
+
+// Stream returns the changing values that this fact has over time.
+// It will first poll sysfs to get the initial cpu count, and then
+// receieves UEvents from the kernel as CPUs are added/removed.
+func (obj CPUCountFact) Stream() error {
+	defer close(obj.init.Output) // Signal when we're done
+
+	ss, err := socketset.NewSocketSet(rtmGrps, socketFile, unix.NETLINK_KOBJECT_UEVENT)
+	if err != nil {
+		return errwrap.Wrapf(err, "error creating socket set")
+	}
+
+	eventChan := make(chan *nlChanEvent) // updated when we receive uevent
+	defer close(eventChan)
+
+	// Start waiting for kernel to poke us about new
+	// device changes on the system
+	go func() {
+		defer ss.Shutdown()
+		defer ss.Close()
+		for {
+			uevent, err := ss.ReceiveUEvent()
+			//fmt.Printf("Sending %s\n", event.Data["SEQNUM"])
+			// TODO: use struct with error field
+			select {
+			case eventChan <- &nlChanEvent{
+				uevent: uevent,
+				err:    err,
+			}:
+			default: // don't block
+			}
+		}
+	}()
+
+	startChan := make(chan struct{})
+	close(startChan) // trigger the first event
+	var cpuCount int64
+	select {
+	case <-startChan:
+		startChan = nil // disable
+		cpuCount, _ = initCPUCount()
+		if err != nil {
+			// TODO: log?
+			cpuCount = 0
+		}
+		obj.init.Output <- &types.IntValue{
+			V: cpuCount,
+		}
+		break
+	case <-obj.closeChan:
+		return nil
+	}
+
+	for {
+		select {
+		case event, ok := <-eventChan:
+			if !ok {
+				continue
+			}
+			if event.err != nil {
+				// TODO: log here instead?
+				return errwrap.Wrapf(event.err, "error receiving uevent")
+			}
+			//fmt.Printf("Receiving %s\n", event.Data["SEQNUM"])
+			cpus, cpuEvent := processUdev(event.uevent)
+			if cpuEvent {
+				cpuCount += cpus
```

Comment:
So it seems you're keeping a running delta AIUI. It might be better to have a function (you already have it) that gets the exact data we want, and then run that every time we get a valid event. Even though the event might produce useful data, we don't necessarily need to use it, unless that sort of information doesn't come from elsewhere.

That makes this more declarative and usually accurate if many events come in all of a sudden or if we loose one somehow and it messes up the count.

Reply:
yeah, this is a really good idea. thanks.